### PR TITLE
Reclaim highlight perf and slow update

### LINKED ISSUE #3419: metal reclaim widget lags the game
### Description

https://i.imgur.com/GLWJlMU.jpeg (imo this isnt a lot of wrecks and should be smooth)

When you have a lot of wrecks on the ground it becomes a lag machine, it seems to recalculate the metal on ground quite often like once per sec, this could probably be changed to say 2-3 secs without impacting the experience, maybe even calculate it once just when you click a reclaimer unit like rezbot instead of constant updates.

Also the way it calculates metal on ground seems too expensive.

Possible lag fix for this? Make the metal calculation more efficient causing less lag?

### Expected Behaviour

Smooth experience.

### Actual Behaviour

Lag hiccups every second, fps drops.

### Reproduction steps

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by efrec**
### Work done

- Use determinant area/shoelace formula for area enclosed by three points, much more efficient than Heron's formula
- Other more minor changes for perf
- Suppressed check for features moving in x, z
- Increase the time between updates to around 3 seconds

#### Addresses Issue(s)

- Partly resolves #3419. This pr focused on making calcs less expensive to perform. It would take more work to make clusters update independently instead the one map-wide update.

### CODE REVIEW

**Comment by drivver44** [comment]
ive given a ping to lexon to review this pr

**Comment by L-e-x-o-n** [comment]
Thanks! I will run some tests in game before additional comments.

For now, one thing I don't like is the increase in check frequency to 3 seconds, I know it improves performance, but it does so at the cost of widget clusters updating slower. This causes delays in updates, widget won't match status on screen for about 3 seconds, this in my opinion is too much delay, feels bad and will be a worse experience for the widget users.

It might make sense to dynamically adjust check frequency based on total cluster count. It should be low and responsive in early to mid game, but some delays should be less noticeable late game with large wreck fields...

I did some benchmarks in the past and by far the most expensive function is the clusterisation one. It used to do it all the time, I added some checks and optimisations to try to limit it to only run when widget is shown on screen and metal changes but it can be further optimised to only clusterise the clusters that changed. In practice, however, clusters are large with many new wrecks and resbots constantly adding/removing wrecks so it is questionable how much performance would be gained effectively.

**Comment by efrec** [comment]
I agree on just about all counts here. I think many of the changes here are freebies but that doesn't lock in the shift to 3s at all. Early on, especially with commander wreckage races, it's not ideal.

I took a first step on incremental clustering but don't see a way to use less CPU overall yet. It may result in a larger tradeoff -- spiky with less CPU, or smooth at higher CPU. If this workload can be scheduled not to coincide with other expensive updates, then imo the current method comes out on top.

**Comment by L-e-x-o-n** [comment]
Is the increase in number size intentional? I think it's too big now...

**Comment by L-e-x-o-n** [comment]
This version also doesn't rotate the reclaim field value numbers when the camera rotates, they never change direction.

File: luaui/Widgets/gui_reclaim_field_highlight.lua [hunk 1]
```diff
@@ -408,7 +413,7 @@ local minFeatureMetal = 9 -- Tick
 local E2M = 1 / 70 -- Converter ratio
 local minDim = 100
 
-local checkFrequency = 30
+local checkFrequency = 2 * Game.gameSpeed
```
Comment (L-e-x-o-n) [inline luaui/Widgets/gui_reclaim_field_highlight.lua:416]: From my in game tests this is still too much. Wrecks are reclaimed or added but values don't update for a full second and reclaim field remains on map, now linking nonexisting ghost wrecks.

**Comment by L-e-x-o-n** [comment]
I know this has already been merged but I think it would be nice if these things were fixed (if unintentional) or potentially changed.
- Numbers not facing camera is a bug,
- Font size being bigger - intentional or not?
- Lower `checkFrequency` might be just my personal preference but I really notice the added delay in widget updating (I also have it always enabled so this makes it more noticeable)

**Comment by efrec** [comment]
I thought my part-2 to this PR would be done sooner. I took a wrong turn and ran out of free time, is all.

The base issue focuses on frame lag when activating the field highlight, so I'll focus on cold starts when I look at this again (soon). Then will look at checkFrequency and incremental updates, which OPTICS should support just based on some reading.

**Pull request merged**